src/docker/shared.rs

```diff
@@ -488,6 +489,271 @@ impl Directories {
+
+    pub fn exists(&self) -> bool {
+        self.exists.load(Ordering::Relaxed)
+    }
+
+    pub fn exists_static() -> bool {
+        // SAFETY: an atomic load.
+        unsafe { CONTAINER.exists() }
+    }
+
+    // when the `docker exec` command finished.
+    pub fn finish(&mut self, is_tty: bool, msg_info: &mut MessageInfo) {
+        // relax the no-timeout and lack of output
+        // ensure we have atomic ordering
+        if self.exists() {
+            let deleter = self.deleter.as_mut().unwrap();
+            if is_tty {
+                deleter.timeout = DEFAULT_TIMEOUT;
+            }
+            deleter.color_choice = msg_info.color_choice;
+            deleter.verbosity = msg_info.verbosity;
+        }
+
+        self.terminate();
+    }
+
+    pub fn finish_static(is_tty: bool, msg_info: &mut MessageInfo) {
+        // SAFETY: internally guarded by an atomic.
+        unsafe {
+            CONTAINER.finish(is_tty, msg_info);
+        }
+    }
+
+    // terminate the container early. leaves the struct in a valid
+    // state, so it's async safe, but so the container will not
+    // be stopped again.
+    pub fn terminate(&mut self) {
+        if self.exists.swap(false, Ordering::SeqCst) {
+            let deleter = self.deleter.as_mut().unwrap();
+            let mut msg_info = MessageInfo::new(deleter.color_choice, deleter.verbosity);
+            container_stop(
+                &deleter.engine,
+                &deleter.name,
+                deleter.timeout,
+                &mut msg_info,
+            )
+            .ok();
+            container_rm(&deleter.engine, &deleter.name, &mut msg_info).ok();
+
+            self.deleter = None;
+        }
+    }
+}
+
+impl Drop for Container {
+    fn drop(&mut self) {
+        self.terminate();
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum ContainerState {
+    Created,
+    Running,
+    Paused,
+    Restarting,
+    Dead,
+    Exited,
+    DoesNotExist,
+}
+
+impl ContainerState {
+    pub fn new(state: &str) -> Result<Self> {
+        match state {
+            "created" => Ok(ContainerState::Created),
+            "running" => Ok(ContainerState::Running),
+            "paused" => Ok(ContainerState::Paused),
+            "restarting" => Ok(ContainerState::Restarting),
+            "dead" => Ok(ContainerState::Dead),
+            "exited" => Ok(ContainerState::Exited),
+            "" => Ok(ContainerState::DoesNotExist),
+            _ => eyre::bail!("unknown container state: got {state}"),
+        }
+    }
+
+    #[must_use]
+    pub fn is_stopped(&self) -> bool {
+        matches!(self, Self::Exited | Self::DoesNotExist)
+    }
+
+    #[must_use]
+    pub fn exists(&self) -> bool {
+        !matches!(self, Self::DoesNotExist)
+    }
+}
+
+// it's unlikely that we ever need to erase a line in the destructors,
+// and it's better than keep global state everywhere, or keeping a ref
+// cell which could have already deleted a line
```

Comment:
this comment doesn't make sense without context of how `cross::shell` works, could be removed or just mention the reason for no `MessageInfo` is because it needs mutability

Reply:
Is the new comment better? I've changed it to:

```rust
// we need to specify drops for the containers, but we
// also need to ensure the drops are called on a
// termination handler. we use an atomic bool to ensure
// that the drop only gets called once, even if we have
// the signal handle invoked multiple times or it fails.
```